vowpalwabbit/core/src/reductions/eigen_memory_tree.cc

```diff
@@ -0,0 +1,787 @@
+{
+  if (cn.examples.size() <= b.leaf_split) { return; }
+
+  std::vector<emt_feats> exs;
+  for (auto& ex : cn.examples) { exs.push_back(ex->base); }
+
+  cn.left = VW::make_unique<emt_node>();
+  cn.right = VW::make_unique<emt_node>();
+  cn.router_weights = emt_router(exs, b.router_type, *b._random_state);
+
+  std::vector<float> projs;
+  for (auto& ex : exs) { projs.push_back(emt_inner(ex, cn.router_weights)); }
+
+  cn.router_decision = emt_median(projs);
+
+  for (auto& ex : cn.examples) { node_route(cn, *ex)->examples.push_back(std::move(ex)); }
+  cn.examples.clear();
+}
+
+void node_insert(emt_node& cn, std::unique_ptr<emt_example> ex)
+{
+  for (auto& cn_ex : cn.examples)
+  {
+    if (cn_ex->full == ex->full) { return; }
+  }
+  cn.examples.push_back(std::move(ex));
+}
+
+emt_example* node_pick(emt_tree& b, single_learner& base, emt_node& cn, const emt_example& ex)
+{
+  if (cn.examples.size() == 0) { return nullptr; }
+
+  float best_score = FLT_MAX;
+  emt_example* best_example = cn.examples[0].get();
+
+  // shuffle the examples to break ties randomly
+  emt_shuffle(cn.examples.begin(), cn.examples.end(), b._random_state.get());
+
+  for (auto const& example : cn.examples)
+  {
+    float score = scorer_predict(b, base, ex, *example);
+
+    if (score < best_score)
+    {
+      best_score = score;
+      best_example = example.get();
+    }
+  }
+
+  return best_example;
+}
+
+void node_predict(emt_tree& b, single_learner& base, emt_node& cn, emt_example& ex, VW::example& ec)
+{
+  auto closest_ex = node_pick(b, base, cn, ex);
+  ec.pred.multiclass = (closest_ex != nullptr) ? closest_ex->label : 0;
+  ec.loss = (ec.l.multi.label != ec.pred.multiclass) ? ec.weight : 0;
+}
+
+void predict(emt_tree& b, single_learner& base, VW::example& ec)
+{
+  b.all->ignore_some_linear = false;
+  emt_example ex(*b.all, &ec);
+
+  emt_node& cn = *tree_route(b, ex);
+  node_predict(b, base, cn, ex, ec);
+  tree_bound(b, &ex);
+}
+
+void learn(emt_tree& b, single_learner& base, VW::example& ec)
+{
+  b.all->ignore_some_linear = false;
+  auto ex = VW::make_unique<emt_example>(*b.all, &ec);
+
+  emt_node& cn = *tree_route(b, *ex);
+  scorer_learn(b, base, cn, *ex, ec.weight);
+  node_predict(b, base, cn, *ex, ec);  // vw learners predict and learn
+
+  tree_bound(b, ex.get());
+  node_insert(cn, std::move(ex));
+  node_split(b, cn);
+}
+
+void end_pass(emt_tree& b)
+{
+  *(b.all->trace_message) << "##### pass_time: " << static_cast<float>(clock() - b.begin) / CLOCKS_PER_SEC << std::endl;
+
+  b.begin = clock();
+}
+/////////////////end of eigen_memory_tree///////////////////
+////////////////////////////////////////////////////////////
+
+///////////////////Save & Load//////////////////////////////////////
+////////////////////////////////////////////////////////////////////
+void save_load_examples(emt_node& n, io_buf& model_file, bool& read, bool& text, std::stringstream& msg)
+{
+  WRITEITVAR(n.examples.size(), "n_examples", n_examples);
```

Comment:
`WRITEITVAR` should not be used. Please use utility functions in `VW::model_utils`